src/Datadog.Trace/AsyncLocalScopeManager.cs

```diff
@@ -8,26 +9,47 @@ internal class AsyncLocalScopeManager : IScopeManager
 
         private readonly AsyncLocalCompat<Scope> _activeScope = new AsyncLocalCompat<Scope>();
 
+        private readonly List<IScopeListener> _scopeListeners = new List<IScopeListener>(); // TODO: This needs to be thread safe...
+
         public Scope Active => _activeScope.Get();
 
         public Scope Activate(Span span, bool finishOnClose)
         {
             var activeScope = _activeScope.Get();
             var scope = new Scope(activeScope, span, this, finishOnClose);
-            _activeScope.Set(scope);
+
             return scope;
         }
 
         public void Close(Scope scope)
         {
-            var current = _activeScope.Get();
+            foreach (IScopeListener listener in _scopeListeners)
+            {
+                listener.AfterScopeClosed();
```

Comment:
Perhaps an event based name like `OnScopeComplete(Scope instance)`
Another thought is that this might be better placed at the end of this method and wrapped in a try.
Two reasons for this, you may alert subscribers that a scope closed and then it might fail, or this might fail, preventing the scope from closing.

Reply:
The reason I suggest passing the scope, is because in a multi-threaded scenario, the "active" instance could be totally different by the time something gets around to dealing with it.

Reply:
Very good points! I've made the changes to pass the scope argument, I'll follow-up on your other recommendations.